polars/polars-io/src/ipc.rs

```diff
@@ -274,6 +273,115 @@ where
     }
 }
 
+#[must_use]
+#[cfg(feature = "partition")]
+/// Write a DataFrame to partitioned Arrow's IPC format
+/// # Example
+///
+/// ```
+/// use polars_core::prelude::*;
+/// use polars_core::df;
+/// use polars_io::ipc::PartitionIpcWriter;
+///
+/// fn example(df: &DataFrame) -> Result<()> {
+///     let df = df!("a" => [1, 1, 2, 3], "b" => [2, 2, 3, 4], "c" => [2, 3, 4, 5]).unwrap();
+///     let by = ["a", "b"];
+///     PartitionIpcWriter::new("./tmp_dir", by)
+///         .finish(&df)
+/// }
+/// ```
+pub struct PartitionIpcWriter<P, I, S> {
+    rootdir: P,
+    by: I,
+    parallel: bool,
+    compression: Option<write::Compression>,
+    phantom: PhantomData<S>,
+}
+
+#[cfg(feature = "partition")]
+impl<P, I, S> PartitionIpcWriter<P, I, S>
+where
+    P: Into<PathBuf>,
+    I: IntoIterator<Item = S>,
+    S: AsRef<str>,
+{
+    pub fn new(rootdir: P, by: I) -> Self {
+        Self {
+            rootdir,
+            by,
+            parallel: true,
+            compression: None,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Set the compression used. Defaults to None.
+    pub fn with_compression(mut self, compression: Option<write::Compression>) -> Self {
+        self.compression = compression;
+        self
+    }
+
+    /// Write the ipc file in parallel (default). The single threaded writer consumes less memory.
+    pub fn write_parallel(mut self, parallel: bool) -> Self {
+        self.parallel = parallel;
+        self
+    }
+
+    pub fn finish(self, df: &DataFrame) -> Result<()> {
+        use polars_core::POOL;
+        use rayon::iter::IntoParallelIterator;
+        use rayon::iter::ParallelIterator;
+        use uuid::Uuid;
+
+        let rootdir: PathBuf = self.rootdir.into();
+        let by: Vec<String> = self.by.into_iter().map(|x| x.as_ref().to_owned()).collect();
+        let group = df.groupby(&by)?;
+
+        if self.parallel {
+            POOL.install(|| {
+                group
+                    .get_groups()
+                    .idx_ref()
+                    .into_par_iter()
+                    .map(|t| {
+                        let mut partition_df =
+                            unsafe { df.take_iter_unchecked(t.1.iter().map(|i| *i as usize)) };
+                        let mut path = resolve_partition_dir(&rootdir, &by, &partition_df);
+                        std::fs::create_dir_all(&path)?;
+
+                        path.push(format!("part-{}.ipc", Uuid::new_v4()));
```

Comment:
Why would we want a UUID here instead of a count?

Reply:
There is no need to use UUID, but some other libraries use it. Using a count makes sense.